Bad, sad error with table data type when inserting duplicate records

**Description:**
<!-- Give a brief description of the issue -->
$title
**Steps to reproduce:**

```ballerina
type Person record {
    int id;
    int age;
    float salary;
    string name;
    boolean married;
};

public function main() {
    testTableAddOnConstrainedTableWithViolation();
}

function testTableAddOnConstrainedTableWithViolation() {
    Person p1 = { id: 1, age: 30, salary: 300.50, name: "jane", married: true };
    Person p2 = { id: 1, age: 30, salary: 300.50, name: "jane", married: true };

    table<Person> t1 = table {
        { key id, salary, name, age, married },
        [p1, p2]
    };
}

```
ballerina-internal.log
```
[2019-08-20 16:56:53,457] ERROR {org.ballerinalang.jvm.scheduling.Scheduler} - Strand died 
org.ballerinalang.jvm.util.exceptions.BallerinaException: execute update failed: Unique index or primary key violation: "PRIMARY KEY ON PUBLIC.TABLE_PERSON_0(ID) [1, 30, 300.5, 'jane', TRUE]"; SQL statement:
INSERT INTO TABLE_PERSON_0 (id ,age ,salary ,name ,married ) values (?,?,?,?,?) [23505-199]
	at org.ballerinalang.jvm.TableUtils.prepareAndExecuteStatement(TableUtils.java:104)
	at org.ballerinalang.jvm.TableProvider.prepareAndExecuteStatement(TableProvider.java:322)
	at org.ballerinalang.jvm.TableProvider.insertData(TableProvider.java:113)
	at org.ballerinalang.jvm.values.TableValue.addData(TableValue.java:244)
	at org.ballerinalang.jvm.values.TableValue.insertInitialData(TableValue.java:370)
	at org.ballerinalang.jvm.values.TableValue.<init>(TableValue.java:111)
	at irshadnilam.examples.main.testTableAddOnConstrainedTableWithViolation(main.bal:17)
	at irshadnilam.examples.main.main(main.bal:10)
	at irshadnilam.examples.___init.$lambda$main$(examples)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:452)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:228)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:198)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: Unique index or primary key violation: "PRIMARY KEY ON PUBLIC.TABLE_PERSON_0(ID) [1, 30, 300.5, 'jane', TRUE]"; SQL statement:
INSERT INTO TABLE_PERSON_0 (id ,age ,salary ,name ,married ) values (?,?,?,?,?) [23505-199]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:457)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:427)
	at org.h2.message.DbException.get(DbException.java:205)
	at org.h2.message.DbException.get(DbException.java:181)
	at org.h2.mvstore.db.MVPrimaryIndex.add(MVPrimaryIndex.java:131)
	at org.h2.mvstore.db.MVTable.addRow(MVTable.java:546)
	at org.h2.command.dml.Insert.insertRows(Insert.java:180)
	at org.h2.command.dml.Insert.update(Insert.java:132)
	at org.h2.command.CommandContainer.update(CommandContainer.java:133)
	at org.h2.command.Command.executeUpdate(Command.java:267)
	at org.h2.jdbc.JdbcPreparedStatement.execute(JdbcPreparedStatement.java:250)
	at org.ballerinalang.jvm.TableUtils.prepareAndExecuteStatement(TableUtils.java:102)
	... 12 more
```
[ballerina-internal.log](https://github.com/ballerina-platform/ballerina-lang/files/3520301/ballerina-internal.log)

**Affected Versions:**
Ballerina 1.0.0-beta-SNAPSHOT
Language specification 2019R2
Ballerina tool 1.0.0

**OS, DB, other environment details and versions:**
MacOs 10.14.5
java version "1.8.0_212"

**Related Issues (optional):**
<!-- Any related issues such as sub tasks, issues reported in other repositories (e.g component repositories), similar problems, etc. -->

**Suggested Labels (optional):**
<!-- Optional comma separated list of suggested labels. Non committers can’t assign labels to issues, so this will help issue creators who are not a committer to suggest possible labels-->

**Suggested Assignees (optional):**
<!--Optional comma separated list of suggested team members who should attend the issue. Non committers can’t assign issues to assignees, so this will help issue creators who are not a committer to suggest possible assignees-->

--- a/bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TableProvider.java
+++ b/bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TableProvider.java
@@ -24,7 +24,6 @@ import org.ballerinalang.jvm.types.BStructureType;
 import org.ballerinalang.jvm.types.BType;
 import org.ballerinalang.jvm.types.BUnionType;
 import org.ballerinalang.jvm.types.TypeTags;
-import org.ballerinalang.jvm.util.exceptions.BallerinaException;
 import org.ballerinalang.jvm.values.ArrayValue;
 import org.ballerinalang.jvm.values.MapValueImpl;
 import org.ballerinalang.jvm.values.TableIterator;
@@ -130,7 +129,7 @@ public class TableProvider {
             itr = new TableIterator(rm, rs, type);
         } catch (SQLException e) {
             releaseResources(conn, stmt);
-            throw new BallerinaException("error in creating iterator for table : " + e.getMessage());
+            throw TableUtils.createTableOperationError("error in creating iterator for table : " + e.getMessage());
         }
         return itr;
     }
@@ -141,7 +140,7 @@ public class TableProvider {
             conn = DriverManager
                     .getConnection(TableConstants.DB_JDBC_URL, TableConstants.DB_USER_NAME, TableConstants.DB_PASSWORD);
         } catch (SQLException e) {
-            throw new BallerinaException("error in getting connection for table db : " + e.getMessage());
+            throw TableUtils.createTableOperationError("error in getting connection for table db : " + e.getMessage());
         }
         return conn;
     }
@@ -169,18 +168,19 @@ public class TableProvider {
                 case TypeTags.UNION_TAG:
                     List<BType> members = ((BUnionType) sf.getFieldType()).getMemberTypes();
                     if (members.size() != 2) {
-                        throw new BallerinaException(UNASSIGNABLE_UNIONTYPE_EXCEPTION);
+                        throw TableUtils.createTableOperationError(UNASSIGNABLE_UNIONTYPE_EXCEPTION);
                     }
                     if (members.get(0).getTag() == TypeTags.NULL_TAG) {
                         generateCreateTableStatement(members.get(1).getTag(), sf, sb);
                     } else if (members.get(1).getTag() == TypeTags.NULL_TAG) {
                         generateCreateTableStatement(members.get(0).getTag(), sf, sb);
                     } else {
-                        throw new BallerinaException(UNASSIGNABLE_UNIONTYPE_EXCEPTION);
+                        throw TableUtils.createTableOperationError(UNASSIGNABLE_UNIONTYPE_EXCEPTION);
                     }
                     break;
                 default:
-                    throw new BallerinaException("Unsupported column type for table : " + sf.getFieldType());
+                    throw TableUtils
+                            .createTableOperationError("Unsupported column type for table : " + sf.getFieldType());
             }
             seperator = ",";
         }
@@ -229,7 +229,8 @@ public class TableProvider {
                 }
                 break;
             default:
-                throw new BallerinaException("Unsupported nillable field for table : " + sf.getFieldType());
+                throw TableUtils
+                        .createTableOperationError("Unsupported nillable field for table : " + sf.getFieldType());
         }
     }
 
@@ -261,7 +262,7 @@ public class TableProvider {
             stmt = conn.createStatement();
             stmt.executeUpdate(queryStatement);
         } catch (SQLException e) {
-            throw new BallerinaException (
+            throw TableUtils.createTableOperationError(
                     "error in executing statement : " + queryStatement + " error:" + e.getMessage());
         } finally {
             releaseResources(conn, stmt);
@@ -307,7 +308,7 @@ public class TableProvider {
             }
             stmt.executeUpdate();
         } catch (SQLException e) {
-            throw new BallerinaException(
+            throw TableUtils.createTableOperationError(
                     "error in executing statement : " + queryStatement + " error:" + e.getMessage());
         } finally {
             releaseResources(conn, stmt);
@@ -321,7 +322,7 @@ public class TableProvider {
             stmt = conn.prepareStatement(queryStatement);
             TableUtils.prepareAndExecuteStatement(stmt, constrainedType);
         } catch (SQLException e) {
-            throw new BallerinaException(
+            throw TableUtils.createTableOperationError(
                     "error in executing statement : " + queryStatement + " error:" + e.getMessage());
         } finally {
             releaseResources(conn, stmt);
@@ -334,14 +335,16 @@ public class TableProvider {
                 stmt.close();
             }
         } catch (SQLException e) {
-            throw new BallerinaException("error in releasing table statement resource : " + e.getMessage());
+            throw TableUtils
+                    .createTableOperationError("error in releasing table statement resource : " + e.getMessage());
         }
         try {
             if (conn != null && !conn.isClosed()) {
                 conn.close();
             }
         } catch (SQLException e) {
-            throw new BallerinaException("error in releasing table connection resource : " + e.getMessage());
+            throw TableUtils
+                    .createTableOperationError("error in releasing table connection resource : " + e.getMessage());
         }
     }
 
@@ -357,8 +360,8 @@ public class TableProvider {
             }
             return rowCount;
         } catch (SQLException e) {
-            throw new BallerinaException("error in executing statement to get the count : " + stmt + " error:"
-                                         + e.getMessage());
+            throw TableUtils.createTableOperationError(
+                    "error in executing statement to get the count : " + stmt + " error:" + e.getMessage());
         } finally {
             releaseResources(conn, stmt);
         }

--- a/bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TableUtils.java
+++ b/bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TableUtils.java
@@ -25,7 +25,6 @@ import org.ballerinalang.jvm.types.BType;
 import org.ballerinalang.jvm.types.BUnionType;
 import org.ballerinalang.jvm.types.TypeTags;
 import org.ballerinalang.jvm.util.exceptions.BallerinaErrorReasons;
-import org.ballerinalang.jvm.util.exceptions.BallerinaException;
 import org.ballerinalang.jvm.values.ArrayValue;
 import org.ballerinalang.jvm.values.ErrorValue;
 import org.ballerinalang.jvm.values.MapValueImpl;
@@ -84,7 +83,7 @@ public class TableUtils {
                     case TypeTags.UNION_TAG:
                         List<BType> members = ((BUnionType) sf.getFieldType()).getMemberTypes();
                         if (members.size() != 2) {
-                            throw new BallerinaException(
+                            throw createTableOperationError(
                                     "Corresponding Union type in the record is not an assignable nillable type");
                         }
                         if (members.get(0).getTag() == TypeTags.NULL_TAG) {
@@ -92,7 +91,7 @@ public class TableUtils {
                         } else if (members.get(1).getTag() == TypeTags.NULL_TAG) {
                             prepareAndExecuteStatement(stmt, data, index, sf, members.get(0).getTag(), fieldName);
                         } else {
-                            throw new BallerinaException(
+                            throw createTableOperationError(
                                     "Corresponding Union type in the record is not an assignable nillable type");
                         }
                         break;
@@ -101,7 +100,7 @@ public class TableUtils {
             }
             stmt.execute();
         } catch (SQLException e) {
-            throw new BallerinaException("execute update failed: " + e.getMessage(), e);
+            throw createTableOperationError("execute update failed: " + e.getMessage());
         }
     }
 
@@ -213,13 +212,18 @@ public class TableUtils {
                 }
                 break;
             default:
-                throw new BallerinaException("unsupported data type for array parameter");
+                throw createTableOperationError("unsupported data type for array parameter");
         }
         return arrayData;
     }
 
     public static ErrorValue createTableOperationError(Throwable throwable) {
         String detail = throwable.getMessage() != null ? throwable.getMessage() : DEFAULT_ERROR_DETAIL_MESSAGE;
+        return BallerinaErrors
+                .createError(BallerinaErrorReasons.TABLE_OPERATION_ERROR, detail);
+    }
+
+    public static ErrorValue createTableOperationError(String detail) {
         return BallerinaErrors
                 .createError(BallerinaErrorReasons.TABLE_OPERATION_ERROR, detail);
     }

--- a/bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/TableValue.java
+++ b/bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/TableValue.java
@@ -224,6 +224,8 @@ public class TableValue implements RefValue, CollectionValue {
         try {
             this.addData(data);
             return null;
+        } catch (ErrorValue e) {
+            return e;
         } catch (Throwable e) {
             return TableUtils.createTableOperationError(e);
         }

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableLiteralSyntaxTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableLiteralSyntaxTest.java
@@ -107,7 +107,7 @@ public class TableLiteralSyntaxTest {
     }
 
     @Test(expectedExceptions = BLangRuntimeException.class,
-          expectedExceptionsMessageRegExp = ".*Unique index or primary key violation:.*", enabled = false)
+          expectedExceptionsMessageRegExp = ".*Unique index or primary key violation:.*")
     public void testTableAddOnConstrainedTableWithViolation() {
         BRunUtil.invoke(result, "testTableAddOnConstrainedTableWithViolation");
     }
@@ -116,6 +116,14 @@ public class TableLiteralSyntaxTest {
     public void testTableAddOnConstrainedTableWithViolation2() {
         BValue[] returns = BRunUtil.invoke(result, "testTableAddOnConstrainedTableWithViolation2");
         Assert.assertTrue((returns[0]).stringValue().contains("Unique index or primary key violation:"));
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp =
+                  ".*execute update failed: Unique index or primary key violation:.*")
+    public void testTableAddOnConstrainedTableWithViolation3() {
+        BRunUtil.invoke(result, "testTableAddOnConstrainedTableWithViolation3");
+        Assert.fail("Expected exception should have been thrown by this point");
     }
 
     @Test

--- a/tests/jballerina-unit-test/src/test/resources/test-src/types/table/table_literal_syntax.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/types/table/table_literal_syntax.bal
@@ -881,6 +881,15 @@ function testRemoveOpAnonymousFilter() returns table<Order> {
     return orderTable;
 }
 
+function testTableAddOnConstrainedTableWithViolation3() {
+    Person p1 = { id: 1, age: 30, salary: 300.50, name: "jane", married: true };
+    Person p2 = { id: 1, age: 30, salary: 300.50, name: "jane", married: true };
+
+    table<Person> t1 = table {
+        { key id, salary, name, age, married },
+        [p1, p2]
+    };
+}
 
 //function testEmptyTableCreate() returns [int, int] {
 //    table<Person> dt3 = table{};

